aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -146,54 +209,180 @@ Transient lifetime services are created each time they're requested. This lifeti
 Scoped lifetime services are created once per request.
 
 > [!WARNING]
-> If you're using a scoped service in a middleware, inject the service into the `Invoke` or `InvokeAsync` method. Don't inject via constructor injection because it forces the service to behave like a singleton.
+> When using a scoped service in a middleware, inject the service into the `Invoke` or `InvokeAsync` method. Don't inject via constructor injection because it forces the service to behave like a singleton. For more information, see [Middleware](xref:fundamentals/middleware/index).
 
 **Singleton**
 
-Singleton lifetime services are created the first time they're requested (or when `ConfigureServices` is run if you specify an instance there) and then every subsequent request will use the same instance. If your application requires singleton behavior, allowing the services container to manage the service's lifetime is recommended instead of implementing the singleton design pattern and managing your object's lifetime in the class yourself.
+Singleton lifetime services are created the first time they're requested (or when `ConfigureServices` is run and an instance is specified with the service registration). Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide user code to manage the object's lifetime in the class.
+
+> [!WARNING]
+> It's dangerous to resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests.
+
+### Constructor injection behavior
+
+Services can be resolved by two mechanisms:
+
+* `IServiceProvider`
+* [ActivatorUtilities](/dotnet/api/microsoft.extensions.dependencyinjection.activatorutilities)
+
+When services are resolved by `IServiceProvider` or `ActivatorUtilities`, constructor injection requires a *public* constructor. Otherwise, the app throws an [InvalidOperationException](/dotnet/api/system.invalidoperationexception):
+
+> A suitable constructor for type '&lt;TYPE&gt;' couldn't be located. Ensure the type is concrete and services are registered for all parameters of a public constructor.
+
+When services are resolved by `ActivatorUtilities`, constructor injection requires that only one applicable constructor exist. Constructor overloads are supported, but only one overload can exist whose arguments can all be fulfilled by dependency injection. If more than one exists, the app throws an `InvalidOperationException`:
+
+> Multiple constructors accepting all given argument types have been found in type '&lt;TYPE&gt;'. There should only be one applicable constructor.
+
+Constructors can accept arguments that aren't provided by dependency injection, but these must support default values.
+
+The following code throws an `InvalidOperationException` because the runtime is unable to resolve the `item` string:
```

Comment:
I think you should remove all that code, just give them the rule. The valid  and invalid code is nonsense. There's no reason to do that. The rule is all you need.

Unless you have reasonable code, don't show it.